onnxruntime/core/optimizer/reshape_fusion.cc

```diff
@@ -0,0 +1,211 @@
+  GraphViewer graph_viewer(graph);
+  const auto& node_topology_list = graph_viewer.GetNodesInTopologicalOrder();
+
+  for (auto node_index : node_topology_list) {
+    auto* p_reshape = graph.GetNode(node_index);
+    if (p_reshape == nullptr)
+      continue;  // we removed the node as part of an earlier fusion
+
+    Node& reshape = *p_reshape;
+    ORT_RETURN_IF_ERROR(Recurse(reshape, modified, graph_level, logger));
+
+    if (!graph_utils::IsSupportedOptypeVersionAndDomain(reshape, "Reshape", {5}) ||
+        !graph_utils::IsSupportedProvider(reshape, GetCompatibleExecutionProviders())) {
+      continue;
+    }
+
+    if (ReshapeFusion::Fuse_Subgraph1(reshape, graph, logger)) {
+      modified = true;
+    }
+  }
+
+  return Status::OK();
+}
+
+/**
+Apply Reshape Fusion. The fowllowing are subgraphs before and after fusion:
+
+Before fusion:
+   [Sub-graph  Root  Node ]
+    |        /            \
+    |    Shape            Shape
+    |       |              |                   (one or two int64[] constant initializers)
+    |    Gather(indice=0)  Gather(indice=1)    a[]        b[] (optional)
+    |       \              /                   /          /
+    |   Unsqueeze      Unsqueeze              /          /
+    |         \        /  ___________________/          /
+    |          \      /  / ____________________________/
+    |           \    /  / /
+     \            Concat
+      \          /
+         Reshape
+
+After fusion:
+    [Sub-graph Root Node]   (Constant Initializers, b is optional)
+                  \         [0, 0, a, b]
+                   \        /
+                    Reshape
+*/
+bool ReshapeFusion::Fuse_Subgraph1(Node& reshape, Graph& graph, const logging::Logger& logger) {
+  const Node* p_root = graph_utils::GetInputNode(reshape, 0);
+
+  const Node* p_concat = graph_utils::GetInputNode(reshape, 1);
+  if (nullptr == p_concat) {
+    return false;
+  }
+  const Node& concat = *p_concat;
+
+  if (!graph_utils::IsSupportedOptypeVersionAndDomain(concat, "Concat", {1, 4})) {
+    return false;
+  }
+
+  auto concat_input_count = concat.InputArgCount().front();
+  if (concat_input_count < 3 || concat_input_count > 4 || concat.GetOutputEdgesCount() > 1) {
+    return false;
+  }
+
+  // path 1: [Root] --> Shape --> Gather(indices=0) --> Unsqueeze (axes=0) --> Concat [input 0]
+  std::vector<graph_utils::MatchEdgeEnd> parent_path{
+      {true, 0, 0, "Unsqueeze", {1}, kOnnxDomain},
+      {true, 0, 0, "Gather", {1}, kOnnxDomain},
+      {true, 0, 0, "Shape", {1}, kOnnxDomain}};
+
+  std::vector<const Node::EdgeEnd*> edges;
+  if (!graph_utils::FindPath(concat, parent_path, edges)) {
+    return false;
+  }
+
+  const Node& unsqueeze_1 = edges[0]->GetNode();
+  const Node& gather_1 = edges[1]->GetNode();
+  const Node& shape_1 = edges[2]->GetNode();
+  if (unsqueeze_1.GetOutputEdgesCount() != 1 || gather_1.GetOutputEdgesCount() != 1 || shape_1.GetOutputEdgesCount() != 1) {
```

Comment:
Can it be that any of these operators have more than one input edges?

Reply:
Since most op does not have optional input, so number of input edges are pre-defined for each operator. There is no need to check number of input edges.